server/src/test/java/org/elasticsearch/cluster/metadata/MetadataDataStreamsServiceTests.java

```diff
@@ -466,6 +479,56 @@ public void testUpdateDataStreamOptions() {
         assertThat(updatedDataStream.getDataStreamOptions(), equalTo(DataStreamOptions.EMPTY));
     }
 
+    public void testDeleteMissing() {
+        DataStream dataStream = DataStreamTestHelper.randomInstance();
+        ProjectState state = ClusterState.builder(ClusterName.DEFAULT).build().projectState(randomProjectIdOrDefault());
```

Comment:
Since we're not adding a project to the cluster state and cluster states by default only have the `default` project, creating/extracting a project state with a random project ID is going to fail (about 50% of the time). The most correct way of doing it would be
```java
final var projectId = randomProjectIdOrDefault();
ProjectState state = ClusterState.builder(ClusterName.Default)
    .putProjectMetadata(ProjectMetadata.builder(projectId))
    .projectState(projectId);
```
That ensures the requested project is always present (but it can still be the default project).

Reply:
Thanks @nielsbauman .  I had to modify slightly:
```java
        ProjectState state = ClusterState.builder(ClusterName.DEFAULT)
            .putProjectMetadata(ProjectMetadata.builder(projectId))
            .build()
            .projectState(projectId);
```
I also changed the MetadataDataStreamsService#deleteDataStreams to throw a `ResourceNotFoundException` if the datastream doesn't exist. (the prior exception was thrown by the `projectFor` code that has since been removed.